x-pack/plugins/endpoint/server/routes/resolver/utils/tree.ts

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { ResolverEvent } from '../../../../common/types';
+import { extractEntityID, extractParentEntityID } from './normalize';
+import { buildPaginationCursor } from './pagination';
+
+interface NodeStats {
+  totalEvents: number;
+  totalAlerts: number;
+}
+
+interface Node {
+  id: string;
+  children: Node[];
+  events: ResolverEvent[];
```

Comment:
Andrew and I talked about this on a zoom call but I'm curious what other people's thoughts are:

Correct me if I'm wrong @andrewstucki but paginating through the related events and alerts in a tree will only be for a specific focus node. My suggestion was that it might make sense to keep that functionality in its own endpoint (that's how it currently is for related events) instead of having the `/tree` endpoint do it as well.

Reply:
So, I'm cool with it either way. The thought was that at first view of resolver you'd already be focused on the node in question, so you'd need to populate the events/alerts anyway, so might as well throw it in the same API response if we're already returning tree structures and wanting to minimize roundtrips to Kibana.

In the meantime, I still left the other APIs around, so we could always just set the query param values for these requests to resolve 0 events or alerts.